buildSrc/src/main/java/gradlebuild/PublishPlugin.java

```diff
@@ -3,31 +3,122 @@
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
 import org.gradle.api.artifacts.dsl.RepositoryHandler;
+import org.gradle.api.plugins.BasePluginConvention;
+import org.gradle.api.publish.Publication;
 import org.gradle.api.publish.PublishingExtension;
+import org.gradle.api.publish.maven.MavenPublication;
+import org.gradle.api.tasks.TaskContainer;
+import org.gradle.api.tasks.TaskProvider;
+import org.gradle.api.tasks.bundling.Jar;
 
+import java.io.File;
+import java.util.concurrent.Callable;
 import java.util.stream.Stream;
 
 public class PublishPlugin implements Plugin<Project> {
+    public static final String LOCAL_FILE_REPOSITORY_NAME = "LocalFile";
+    private static final String PUBLISH_USER_NAME_PROPERTY = "publishUserName";
+    private static final String PUBLISH_API_KEY_PROPERTY = "publishApiKey";
 
     @Override
     public void apply(Project project) {
-        project.getPlugins().apply(BasePublishPlugin.class);
-        BintrayCredentials credentials = project.getExtensions().getByType(BintrayCredentials.class);
+        project.getPlugins().apply("maven-publish");
 
+        configureVariants(project);
+        configureMainPublication(project);
+        configureRepositories(project);
+    }
+
+    private void configureVariants(Project project) {
+        VariantsExtension variants = project.getExtensions().create("variants", VariantsExtension.class);
+        variants.getGroupId().convention(project.provider(() -> project.getGroup().toString()));
+        variants.getArtifactId().convention(project.provider(() -> getArchivesBaseName(project)));
+        variants.getVersion().convention(project.provider(() -> project.getVersion().toString()));
+    }
+
+    private void configureMainPublication(Project project) {
+        TaskContainer tasks = project.getTasks();
+        TaskProvider<Jar> jarTask = project.getTasks().named("jar", Jar.class);
+        TaskProvider<Jar> sourceZipTask = tasks.register("sourceZip", Jar.class, jar -> jar.getArchiveClassifier().set("sources"));
+        TaskProvider<Jar> javadocZipTask = tasks.register("javadocZip", Jar.class, jar -> jar.getArchiveClassifier().set("javadoc"));
+        project.getExtensions().configure(
+            PublishingExtension.class,
+            extension -> extension.getPublications().create("main", MavenPublication.class, main -> {
+                main.artifact(jarTask.get());
+                main.artifact(sourceZipTask.get());
+                main.artifact(javadocZipTask.get());
```

Comment:
Do you need to call `get()` here? Can't we just pass the provider?

Reply:
Looks like the provider is supported starting from Gradle 6.6: https://github.com/gradle/gradle/pull/13505

Let me try to remove the `get` now.

Reply:
Removed the `get()`.